dev/NumberBox/NumberBox.cpp

```diff
@@ -0,0 +1,387 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License. See LICENSE in the project root for license information.
+
+#include "pch.h"
+#include "common.h"
+#include "NumberBox.h"
+#include "NumberBoxAutomationPeer.h"
+#include "RuntimeProfiler.h"
+#include "ResourceAccessor.h"
+#include "Utils.h"
+
+NumberBox::NumberBox()
+{
+    __RP_Marker_ClassById(RuntimeProfiler::ProfId_NumberBox);
+
+    // Default values for the number formatter
+    auto formatter = winrt::DecimalFormatter();
+    formatter.IntegerDigits(1);
+    formatter.FractionDigits(0);
+    NumberFormatter(formatter);
+
+    PointerWheelChanged({ this, &NumberBox::OnScroll });
+
+    SetDefaultStyleKey(this);
+}
+
+winrt::AutomationPeer NumberBox::OnCreateAutomationPeer()
+{
+    return winrt::make<NumberBoxAutomationPeer>(*this);
+}
+
+void NumberBox::OnApplyTemplate()
+{
+    winrt::IControlProtected controlProtected = *this;
+
+    if (auto spinDown = GetTemplateChildT<winrt::RepeatButton>(L"DownSpinButton", controlProtected))
+    {
+        spinDown.Click({ this, &NumberBox::OnSpinDownClick });
+
+        // Do localization for the down button
+        if (winrt::AutomationProperties::GetName(spinDown).empty())
+        {
+            auto spinDownName = ResourceAccessor::GetLocalizedStringResource(SR_NumberBoxDownSpinButtonName);
+            winrt::AutomationProperties::SetName(spinDown, spinDownName);
+        }
+    }
+
+    if (auto spinUp = GetTemplateChildT<winrt::RepeatButton>(L"UpSpinButton", controlProtected))
+    {
+        spinUp.Click({ this, &NumberBox::OnSpinUpClick });
+
+        // Do localization for the up button
+        if (winrt::AutomationProperties::GetName(spinUp).empty())
+        {
+            auto spinUpName = ResourceAccessor::GetLocalizedStringResource(SR_NumberBoxUpSpinButtonName);
+            winrt::AutomationProperties::SetName(spinUp, spinUpName);
+        }
+    }
+
+    m_textBox.set([this, controlProtected]() {
```

Comment:
>m_textBox.set([this, controlProtected]() { [](start = 4, length = 42)

I think I have told you to do something like this in the past for scope guarding benefits, I think you can get all the same scope guards with cleaner looking code here by using the if statement defined local.

```
if(const auto textBox = GetTemplateChildT<winrt::TextBox>(L"InputBox", controlProtected)
{
    textBox.LostFocus({ this, &NumberBox::OnTextBoxLostFocus });
    textBox.KeyUp({ this, &NumberBox::OnNumberBoxKeyUp });
    m_textBox.set(textBox);
}
```

Reply:
In that case, though, if "InputBox" can't be found, the value of m_textBox doesn't get set, so it could be left pointing to some old UI that doesn't exist anymore. What I used to do was:

```
m_textBox.set(GetTemplateChildT<winrt::TextBox>(c_tabViewTextBoxName, controlProtected));
if (auto textBox = m_textBox.get())
{
    textBox.LostFocus({ this, &NumberBox::OnTextBoxLostFocus });
    textBox.KeyUp({ this, &NumberBox::OnNumberBoxKeyUp });
}
```

Do you prefer that or the lambda thing?

Reply:
hmm, yeah I didn't think about OnApplyTemplate being called a second time.. I probably prefer the lambda then